core/DataAccess/LogAggregator.php

```diff
@@ -174,18 +187,107 @@ public function setQueryOriginHint($nameOfOrigiin)
     {
+        $segment = $this->segment;
         $bind = $this->getGeneralQueryBindParams();
-        $query = $this->segment->getSelectQuery($select, $from, $where, $bind, $orderBy, $groupBy, $limit, $offset);
+
+        if (!$this->segment->isEmpty() && $this->isSegmentCacheEnabled()) {
+            // todo when we create table for query we cannot know if the archiver will use that same Db writer or Db reader
+            // if plugin doesn't support reader just yet... there will be a problem!
+            $segment = new Segment('', $this->sites);
+
+            $segmentTable = $this->getSegmentTmpTableName();
+
+            $segmentWhere = $this->getWhereStatement('log_visit', 'visit_last_action_time');
+            $segmentBind = $this->getGeneralQueryBindParams();
+
+            $logQueryBuilder = StaticContainer::get('Piwik\DataAccess\LogQueryBuilder');
+            $forceGroupByBackup = $logQueryBuilder->getForcedInnerGroupBySubselect();
+            $logQueryBuilder->forceInnerGroupBySubselect('');
+            $segmentSql = $this->segment->getSelectQuery('distinct log_visit.idvisit as idvisit', 'log_visit', $segmentWhere, $segmentBind, 'log_visit.idvisit ASC');
+            $logQueryBuilder->forceInnerGroupBySubselect($forceGroupByBackup);
+            
+            try {
+                Db::getReader()->query('CREATE TEMPORARY TABLE IF NOT EXISTS ' . Common::prefixTable($segmentTable) . ' (idvisit  BIGINT(10) UNSIGNED NOT NULL) ' . $segmentSql['sql'], $segmentSql['bind']);
+
+                if (!is_array($from)) {
+                    $from = array($segmentTable, $from);
+                } else {
+                    array_unshift($from, $segmentTable);
+                }
+
+                $logTablesProvider = $this->getLogTableProvider();
+                $logTablesProvider->setTempTable(new LogTableTemporary($segmentTable));
+
+                foreach ($logTablesProvider->getAllLogTables() as $logTable) {
+                    if ($logTable->getDateTimeColumn()) {
+                        $whereTest = $this->getWhereStatement($logTable->getName(), $logTable->getDateTimeColumn());
+                        if (strpos($where, $whereTest) === 0) {
+                            // we don't need to apply the where statement again as it would have been applied already
+                            // in the temporary table... instead it should join the tables through the idvisit index
+                            $where = ltrim(str_replace($whereTest, '', $where));
+                            if (stripos($where, 'and ') === 0) {
+                                $where = substr($where, strlen('and '));
+                            }
+                            $bind = array_slice($bind, 3);
+                            break;
+                        }
+                    }
+
+                }
+            } catch (\Exception $e) {
+                throw $e;
+            }
```

Comment:
Given it just rethrows, is this try-catch necessary?

Reply:
Same for below.

Reply:
Will remove it again 👍 had it in there for easier debugging